text/0082-entity-tags.md

```diff
@@ -0,0 +1,251 @@
+
+Capability tracking must be generalized to support tags. In particular, `tags T` should be treated as a record with optional attributes, and `hasTag` checks on its keys should update the capability set. For our introductory example, consider the following expression
+
+```
+  resource.hasTag("write") && // (1)
+  principal.hasTag("write") &&  // (2)
+  resource.getTag("write").containsAny(principal.getTag("write")) // (3)
+```
+
+After the subexpression (1), `resource.hasTag("write")` should be added to the current capability set. After subexpression (2), `principal.hasTag("write")` should be added to it. Finally, when validating subexpression (3), the expression `resource.getTag("write")` will be considered valid since `resource.hasTag("write")` is in the current capability set and it will be given type `Set<String>`, as the `tags` for `resource` has type `Set<String>`. The expression `principal.getTag("write")` is handled similarly. If either of the `hasTag` subexpressions (1) or (2) were omitted, subexpression (3) would not validate due to the missing capability set entries.
+
+For entity types with no `tags` declaration in the schema, the validator gives `hasTag` the type `False` (to support short-circuiting), and rejects `getTag`.
+
+
+### Validating and parsing entities
+
+The Cedar authorizer's `is_authorized` function can be asked to validate that entities in a request are consistent with a provided schema. Extending validation to work with tags is straightforward. We check that every value in the `tags` map for an entity `e` has type `T` given that type of `e` is declared to have `tags T`.  If the type `e` doesn't have `tags` in the schema, then specifying a `tags` field is an error.
+
+Similarly, schema-based parsing considers schemas when parsing in entities, and it can confirm when parsing that the contents of `tags` attribute in the JSON have the appropriate shape, and if the type `T` has different behavior under schema-based parsing (for instance, if it is an extension type), then schema-based parsing can interpret the `tags` appropriately.
+
+
+### Symbolic compilation
+
+Tags as defined here have an efficient logical representation as *uninterpreted binary functions*.  In particular, for each declaration `entity E ... tags T`, we introduce an uninterpreted function `f_E` of type `E -> String -> Option T`.  With this representation, we can straightforwardly translate the `hasTag` and `getTag` operations as applications of the function `f_E`, where the absence of a key is denoted by binding it to the value `(none T)`.
+
+
+## Drawbacks
+
+Entity types are permitted to have a single `tags` declaration `tags T`, which eliminates the possibility of attaching multiple tags-maps to a single entity (unlike [RFC 68](https://github.com/cedar-policy/rfcs/pull/68)).  This RFC also does not support tags containing other tags-maps, comparing tags-maps with `==`, and the like (as discussed above).
```

Comment:
Does this forever limit you to a single set of key/value types to authorize on? Ever Kubernetes object has string key/value `labels` _and_ `annotations`. I don't think most users make authorization decisions based on `annotation` values, but some might want policy to require a specific `annotation` key has a value for audit purposes. Could I make a separate set of tags a named attribute on an entity?

Reply:
Yes, this is possible.  Paragraphs below this one show how to get arbitrarily many tag-maps attached to a single entity, and also how to achieve (if desired) arbitrary nesting of tag-maps into other types.

Reply:
ah ok I get it now, thank you!